src/VisualStudio/Core/Def/Implementation/Remote/RemoteHostClientServiceFactory.SolutionChecksumUpdator.cs

```diff
@@ -128,21 +133,40 @@ private void OnWorkspaceChanged(object sender, WorkspaceChangeEventArgs e)
 
             private async Task UpdateSolutionChecksumAsync(CancellationToken cancellationToken)
             {
-                // hold onto previous snapshot
-                var previousSnapshot = _lastSnapshot;
+                using (await _gate.DisposableWaitAsync(cancellationToken).ConfigureAwait(false))
+                {
+                    // hold onto previous snapshot
+                    var previousSnapshot = _lastSnapshot;
 
-                // create a new one (incrementally update the snapshot)
-                _lastSnapshot = await _checksumService.CreateChecksumAsync(_workspace.CurrentSolution, cancellationToken).ConfigureAwait(false);
+                    // create a new one (incrementally update the snapshot)
+                    _lastSnapshot = await _checksumService.CreateChecksumAsync(_service.Workspace.CurrentSolution, cancellationToken).ConfigureAwait(false);
```

Comment:
since it returns disposable object, I didn't want to deal with race. if I move it out from lock, I need queue that serialize work to set _lastSnapshot with proper dispose of other snapshot.